airflow/ti_deps/deps/trigger_rule_dep.py

```diff
@@ -117,70 +229,120 @@ def _evaluate_trigger_rule(
-            except (NotFullyPopulated, NotMapped):
-                return None
-            return ti.get_relevant_upstream_map_indexes(
-                upstream_tasks[upstream_id],
-                expanded_ti_count,
-                session=session,
-            )
+        new_state = None
+        changed = False
 
-        def _is_relevant_upstream(upstream: TaskInstance) -> bool:
-            """Whether a task instance is a "relevant upstream" of the current task."""
-            # Not actually an upstream task.
-            if upstream.task_id not in task.upstream_task_ids:
-                return False
-            # The current task is not in a mapped task group. All tis from an
-            # upstream task are relevant.
-            if task.get_closest_mapped_task_group() is None:
-                return True
-            # The upstream ti is not expanded. The upstream may be mapped or
-            # not, but the ti is relevant either way.
-            if upstream.map_index < 0:
-                return True
-            # Now we need to perform fine-grained check on whether this specific
-            # upstream ti's map index is relevant.
-            relevant = _get_relevant_upstream_map_indexes(upstream.task_id)
-            if relevant is None:
-                return True
-            if relevant == upstream.map_index:
-                return True
-            if isinstance(relevant, collections.abc.Container) and upstream.map_index in relevant:
-                return True
-            return False
+        # if there's a failure, we mark upstream_failed; if there's a skip, we mark skipped
+        # in either case, we don't wait for all relevant setups to complete
+        if dep_context.flag_upstream_failed:
+            if upstream_failed or failed:
+                new_state = TaskInstanceState.UPSTREAM_FAILED
+            elif skipped:
+                new_state = TaskInstanceState.SKIPPED
+            elif removed and success and ti.map_index > -1:
+                if ti.map_index >= success:
+                    new_state = TaskInstanceState.REMOVED
+
+        if new_state is not None:
+            if new_state == TaskInstanceState.SKIPPED and dep_context.wait_for_past_depends_before_skipping:
+                past_depends_met = ti.xcom_pull(
+                    task_ids=ti.task_id, key=PAST_DEPENDS_MET, session=session, default=False
+                )
+                if not past_depends_met:
+                    yield self._failing_status(
+                        reason="Task should be skipped but the past depends are not met"
+                    ), changed
+                    return
+            changed = ti.set_state(new_state, session)
+
+        if changed:
+            dep_context.have_changed_ti_states = True
+
+        non_successes = upstream - success
+        if ti.map_index > -1:
+            non_successes -= removed
+        if non_successes > 0:
```

Comment:
I think we can replace all of this by `done` after adding it:
```suggestion
        if not done:
```

Reply:
this could be true but, this logic was added to deal with a deadlock and i don't really want to mess with it here. i'm just keeping the behavior consistent with trigger rule. maybe let's do this in followup?